src/field_theory/galois.lean

```diff
@@ -0,0 +1,233 @@
+  mul_left_inv := λ ϕ, by {ext, exact symm_apply_apply ϕ a} }
+
+lemma is_galois_implies_card_aut_eq_findim [finite_dimensional F E] [h : is_galois F E] :
+  fintype.card (E ≃ₐ[F] E) = findim F E :=
+begin
+  cases field.exists_primitive_element h.1 with α hα,
+  cases h.1 α with H1 h_separable,
+  cases h.2 α with H2 h_splits,
+  have switch : (⊤ : intermediate_field F E).to_subalgebra.to_submodule = ⊤ :=
+    by { ext, exact iff_of_true intermediate_field.mem_top submodule.mem_top },
+  rw [←findim_top, ←switch],
+  change fintype.card (E ≃ₐ[F] E) = findim F (⊤ : intermediate_field F E),
+  replace h_splits : polynomial.splits (algebra_map F F⟮α⟯) (minimal_polynomial H2),
+  { rw hα,
+    let map : E →+* (⊤ : intermediate_field F E) :=
+    { to_fun := λ x, ⟨x, intermediate_field.mem_top⟩,
+      map_one' := rfl,
+      map_mul' := λ _ _, rfl,
+      map_zero' := rfl,
+      map_add' := λ _ _, rfl },
+    rw (show algebra_map F (⊤ : intermediate_field F E) = map.comp (algebra_map F E),
+      by { ext, refl }),
+    exact polynomial.splits_comp_of_splits (algebra_map F E) map h_splits },
+  rw [←hα, intermediate_field.adjoin.findim H2],
+  rw ← intermediate_field.card_alg_hom_adjoin_integral F H2 h_separable h_splits,
+  apply fintype.card_congr,
+  transitivity (F⟮α⟯ ≃ₐ[F] F⟮α⟯),
+  { rw hα,
+    change (E ≃ₐ[F] E) ≃ ((⊤ : intermediate_field F E).to_subalgebra ≃ₐ[F]
+      (⊤ : intermediate_field F E).to_subalgebra),
+    rw intermediate_field.top_to_subalgebra,
+    exact
+    { to_fun := λ ϕ, (algebra.top_equiv).trans (trans ϕ (algebra.top_equiv).symm),
+      inv_fun := λ ϕ, (algebra.top_equiv).symm.trans (trans ϕ (algebra.top_equiv)),
+      left_inv := λ _, by { ext, simp only [apply_symm_apply, trans_apply] },
+      right_inv := λ _, by { ext, simp only [symm_apply_apply, trans_apply] } } },
+  { exact alg_equiv_equiv_alg_hom F F⟮α⟯ },
+end
+
+end
+
+section galois_correspondence
+
+variables {F : Type*} [field F] {E : Type*} [field E] [algebra F E]
+variables (H : subgroup (E ≃ₐ[F] E)) (K : intermediate_field F E)
+
+instance tower_top_of_galois [h : is_galois F E] : is_galois K E :=
+⟨is_separable_tower_top_of_is_separable K h.1, normal.tower_top_of_normal F K E h.2⟩
+
+instance subgroup_action : faithful_mul_semiring_action H E :=
+{ smul := λ h x, h x,
+  smul_zero := λ _, map_zero _,
+  smul_add := λ _, map_add _,
+  one_smul := λ _, rfl,
+  smul_one := λ _, map_one _,
+  mul_smul := λ _ _ _, rfl,
+  smul_mul := λ _, map_mul _,
+  eq_of_smul_eq_smul' := λ x y z, subtype.ext (ext z) }
+
+/-- The intermediate_field fixed by a subgroup -/
+def fixed_field : intermediate_field F E :=
+{ carrier := mul_action.fixed_points H E,
+  zero_mem' := smul_zero,
+  add_mem' := λ _ _ hx hy _, by rw [smul_add, hx, hy],
+  neg_mem' := λ _ hx _, by rw [smul_neg, hx],
+  one_mem' := smul_one,
+  mul_mem' := λ _ _ hx hy _, by rw [smul_mul', hx, hy],
+  inv_mem' := λ _ hx _, by rw [smul_inv, hx],
+  algebra_map_mem' := λ _ _, commutes _ _ }
```

Comment:
There is some really slow elaboration going on here, I suspect typeclass search since it generates a large amount of output. Making arguments explicit helps a lot, but not enough (from 6.5s to 1.8s on my machine). Maybe the correct option is to make the automorphism group a structure or `irreducible`.

```suggestion
def fixed_field : intermediate_field F E :=
{ carrier := mul_action.fixed_points H E,
  zero_mem' := λ g, smul_zero g,
  add_mem' := λ a b hx hy g, by rw [smul_add g a b, hx, hy],
  neg_mem' := λ a hx g, by rw [smul_neg g a, hx],
  one_mem' := λ g, smul_one g,
  mul_mem' := λ a b hx hy g, by rw [smul_mul' g a b, hx, hy],
  inv_mem' := λ a hx g, by rw [smul_inv _ g a, hx],
  algebra_map_mem' := λ a g, commutes g a }
```